tools/flight-visualizer/src/data.rs

```diff
@@ -0,0 +1,231 @@
+const MAX_DATA_POINTS: usize = 10;
+
+pub struct Data {
+    pub time: f64,
+    pub bme: BmeData,
+    pub orientation: Option<Orientation>,
+    pub acceleration: Option<Acceleration>,
+    pub signal_strength: Option<SignalStrength>,
+}
+
+pub struct BmeData {
+    pub temperature: OffsetVec<[f64; 2]>,
+    pub pressure: OffsetVec<[f64; 2]>,
+    pub height: OffsetVec<[f64; 2]>,
+}
+
+pub struct SignalStrength {
+    pub rssi: i32,
+    pub snr: i32,
+}
+
+pub struct Orientation {
+    pub x: f32,
+    pub y: f32,
+    pub z: f32,
+}
+
+pub struct Acceleration {
+    pub x: f32,
+    pub y: f32,
+    pub z: f32,
+}
+
+impl Data {
+    pub fn empty() -> Self {
+        Self {
+            time: 0f64,
+            bme: BmeData::empty(),
+            orientation: None,
+            acceleration: None,
+            signal_strength: None,
+        }
+    }
+}
+
+impl BmeData {
+    pub fn empty() -> Self {
+        Self {
+            temperature: OffsetVec::new(MAX_DATA_POINTS),
+            pressure: OffsetVec::new(MAX_DATA_POINTS),
+            height: OffsetVec::new(MAX_DATA_POINTS),
+        }
+    }
+}
+
+pub struct OffsetVec<T> {
+    vec: Vec<T>,
+    offset: usize,
+    max_size: usize,
+    elem_count: usize,
+}
+
+impl<T> OffsetVec<T>
+where
+    T: Clone + Copy,
+{
+    pub fn new(max_size: usize) -> Self {
+        OffsetVec {
+            vec: Vec::with_capacity(max_size),
+            offset: 0,
+            elem_count: 0,
+            max_size,
+        }
+    }
+
+    pub fn push(&mut self, elem: T) {
+        if self.elem_count < self.max_size {
+            self.elem_count += 1;
+            self.vec.push(elem);
+        } else {
+            self.vec[self.offset] = elem;
+        }
+        self.offset = (self.offset + 1) % self.max_size;
+    }
+
+    pub fn get(&self, index: usize) -> Option<&T> {
+        if index >= self.elem_count {
+            return None;
+        }
+        if self.elem_count < self.max_size {
+            return Some(&self.vec[index]);
+        }
+        let index = (self.offset + index) % self.max_size;
+
+        Some(&self.vec[index])
+    }
+
+    pub fn to_vec(&self) -> Vec<T> {
+        //TODO optimize
+        let mut vec: Vec<T> = Vec::with_capacity(self.elem_count);
+        for i in 0..self.elem_count {
+            vec.push(*self.get(i).unwrap());
+        }
+
+        vec
+    }
+
+    pub fn get_elem_count(&self) -> usize {
+        self.elem_count
+    }
+
+    pub fn get_max_size(&self) -> usize {
+        self.max_size
+    }
+}
+
+pub struct OffsetVecIter<'a, T>
```

Comment:
just use a [`std::collections::VecDeque`](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html) (pronounced deck)

```rust
if deque.len() == max_len {
    deque.pop_front();
}
deque.push_back(value);
```
you can wrap it in a `CircularBuffer` newtype if you want (and reuse VecDeque iterators), but it may be simpler if you simply don't.

Each time I inspected a collection data structure implemented at work I concluded that It was poorly designed and implemented, it introduced a ton of undocumented code that needed maintenance and we simply didn't need it at all. I'd strongly recommend avoiding writing them unless you know you really need them and you know what you are doing. It's really hardy to get those things right.

Reply:
done